### PR TITLE
[ nG transformation ] Const -> FQ -> Reshape fuse

### PR DESCRIPTION
**Pull request by jane-intel**
**Ticket:** 39124

Transformation logic explanation:
Reshape-Broadcast-Fusion

Required by the "new" form of quantized weights for depthwise convolution from TF:
image

IR for such a sub-graph would look like: 
image

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ngraph::pass::FakeQuantizeReshapeFusion::FakeQuantizeReshapeFusion() {
+    const auto fq_node_p = ngraph::pattern::wrap_type<opset4::FakeQuantize>(
+            {ngraph::pattern::wrap_type<opset4::Constant>(), // for weights only
+             ngraph::pattern::any_input(),
+             ngraph::pattern::any_input(),
+             ngraph::pattern::any_input(),
+             ngraph::pattern::any_input()},
+            pattern::consumers_count(1));
+    const auto reshape_node_p = ngraph::pattern::wrap_type<opset4::Reshape>(
+            {fq_node_p, ngraph::pattern::any_input()}, pattern::consumers_count(1));
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp:24]: Why you check consumers here? Is it really needed?

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher &m) {
+        const auto &pattern_map = m.get_pattern_value_map();
+        const auto fq_node = pattern_map.at(fq_node_p).get_node_shared_ptr();
+        if (fq_node->is_dynamic())
+            return false;
+        const auto &reshape_node = pattern_map.at(reshape_node_p).get_node_shared_ptr();
+        const auto &reshape_pattern_node = reshape_node->get_input_node_shared_ptr(1);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp:32]: You use reshape_pattern_node only as input for new node. So its better to use Output<Node> instead of Node pointer.
```suggestion
        const auto &reshape_pattern_node = reshape_node->input_value(1);
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        const auto &original_data_rank = fq_node->get_input_shape(0).size();
+        OutputVector renewed_inputs = {reshape_node->clone_with_new_inputs({fq_node->input_value(0), reshape_pattern_node})};
+        for (auto i = 1; i < 5; ++i) {
+            Output<Node> limit_input = fq_node->input_value(i);
+            auto limit_shape = limit_input.get_shape();
+            NGRAPH_CHECK(limit_shape.size() <= original_data_rank, "FakeQuantize limit input has unexpected rank");
+            if (limit_shape.size() < original_data_rank) // aligning limit rank with data rank
+                limit_shape.insert(limit_shape.begin(), original_data_rank - limit_shape.size(), uint64_t(1));
+            NGRAPH_CHECK(limit_shape.size() == original_data_rank, "FakeQuantize limit input has unexpected rank");
+            const auto &limit_size = std::accumulate(limit_shape.begin(), limit_shape.end(), 1, std::multiplies<int64_t>());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp:42]:
```suggestion
            const auto &limit_size = shape_size(limit_shape);
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            const auto &max_element_idx = std::distance(limit_shape.begin(), std::max_element(limit_shape.begin(), limit_shape.end()));
+            const auto &max_element = limit_shape[max_element_idx];
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp:44]: You can remove line above and replace current with:
```suggestion
            const auto &max_element = *std::max_element(limit_shape.begin(), limit_shape.end());
```

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            if (max_element == limit_size) { // per-tensor / per-channel limit
+                auto new_limit_shape = reshape_node->get_output_shape(0);
+                std::transform(new_limit_shape.begin(), new_limit_shape.end(), new_limit_shape.begin(),
+                               [max_element](size_t &dim) { return dim == max_element ? max_element : 1; });
+                const auto &new_limit_size = std::accumulate(new_limit_shape.begin(), new_limit_shape.end(), 1, std::multiplies<size_t>());
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp:49]:
```suggestion
                const auto &new_limit_size = shape_size(new_limit_shape);
```

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp (hunk 1 above)
Comment (jane-intel) [reply inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp]: Nope. Copy-paste problems, sorry.

**Review by GlebKazantaev — APPROVED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            return false;
+        }
+        for (auto &new_input : renewed_inputs)
+            copy_runtime_info({reshape_node, fq_node}, new_input.get_node_shared_ptr());
+        const auto new_fq_node = fq_node->clone_with_new_inputs(renewed_inputs);
+        replace_node(reshape_node, new_fq_node);
+        new_fq_node->set_friendly_name(fq_node->get_friendly_name());
+        copy_runtime_info({fq_node, reshape_node}, new_fq_node);
+        return true;
+    };
```
Comment (slyalin) [inline inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp:66]: It's hard to follow without more descriptive comments.

**Comment by jane-intel** [comment]
With informal agreement from @slyalin I merge this PR to master branch.
